### PR TITLE
Make samples_per_epoch optional

### PR DESCRIPTION
**Pull request by carlthome**
For datasets that are procedurally generated and cannot fit onto disk it would be nice to not have to count how many samples there are because generating all data for one epoch can in itself take several hours. For example, when prototyping different data augmentation strategies it would be nice to be able to immediately start training while generating the data.

I've made initial changes to `fit_generator()` to accommodate this. **By setting `samples_per_epoch=None` an epoch can be manually ended by yielding `None` in the generator.** Everything else is as normal. 

1. A downside is of course that the ProgressBar cannot count ETA and loss because it doesn't know how many samples there will be, but this should only matter for the first epoch, after which it would be trivial to get all status reporting up and running again for the following epochs. I'd fix that before an accept of the pr, of course.
2. Another minor issue is that the user is expected to maintain the same number of samples for all epochs, but could input a generator with variable number of samples per epoch, which although pretty cool probably wouldn't make much sense in most applications, but who knows.

**Worthwhile addition?** If so I'll make sure that `evaluate_generator()` and `predict_generator()` work the same way before a merge.
# Example:
```Python
data = load_data()  # Original dataset that fits into RAM.
def augment_data(data):
    # Transform data to increase its comprehensiveness
    yield data
    yield f(data)
    yield g(data)
    # and so on... where f() and g() are arbitrary functions.
def data_generator():
    while True:
        for data in augment_data(data):
            for batch in data:
                yield batch
        yield None
```

### CODE REVIEW

File: keras/engine/training.py [hunk 1]
```diff
@@ -642,8 +642,6 @@ def compile(self, optimizer, loss, metrics=[], loss_weights=None,
         self.predict_function = None
 
     def _make_train_function(self):
-        if not hasattr(self, 'train_function'):
```
Comment (carlthome) [inline keras/engine/training.py]: Some recent changes were lost. I'll fix this.

**Comment by joelthchao** [comment]
`EarlyStopping` callback may help you control the training flow? 
Also, inserting some other stopping flag like `self.model.stop_training` in `_fit_loop` might help you achieve your purpose without changing codes so much.

**Comment by carlthome** [comment]
No, that would not work. 

`EarlyStopping` does not help, and neither does the `stop_training` flag. I'm concerned with being able to stop a single epoch but still continuing the model's training.

Unless you propose running restarting `fit_generator` on every epoch and checking `stop_training` after each batch, but that would screw up the batch generation. The extra thread needs to keep running over several epochs so at least a couple of batches can be buffered up in the `generator_queue` at all times, except for the first epoch of course in which some warm-up is needed because the queue would be empty at first.

**Comment by fchollet** [comment]
> For datasets that are procedurally generated and cannot fit onto disk it would be nice to not have to count how many samples there are because generating all data for one epoch can in itself take several hours. 

If you are generating them procedurally, surely you know how many will be generated?

**Comment by fchollet** [comment]
Overall I think the PR clashes with the way `fit_generator` is meant to be used. The general principle is that your generators should yield forever (either an infinite loop or perpetual synthetic augmentations), and that `samples_per_epoch` determines how many samples the model should draw from the never-ending generator at each epoch.

**Comment by mguillau** [comment]
Sorry to reopen this PR, but it is quite related to a simple PR I was thinking of submitting.

Would you consider favorably the possibility for the generator to provide the {fit,evaluate,predict}_generator() functions its number of samples per epoch? The generator object would just need to have the 'samples_per_epoch' attribute, which would be used if it is not explicitly provided to the function.

(This attribute is something that I already use for my generators and it would save me a few characters -- _il n'y a pas de petit profits_ -- in calls such as: `model.fit_generator( traingenerator, samples_per_epoch=traingenerator.samples_per_epoch, ... )` )

**Comment by carlthome** [comment]
@mguillau, I'm not particularly familiar with Python but couldn't you do something like:

```Python
class A:
  __init__(self):
    self.samples_per_epoch = ...  # Whatever value

  __call__(self):
      # Generator stuff here
```

**Comment by carlthome** [comment]
@fchollet, 
> If you are generating them procedurally, surely you know how many will be generated?
No. I'm doing audio analysis of songs, and each song (i.e. a sequence of features) needs to be split and padded with silence, but how much depends on how the songs were ordered into batches. If every epoch is shuffled differently the samples_per_epoch differs (which is fine because it's masked out in the model). This is currently impossible to do in Keras, so one has to settle for the same ordering of songs throughout training, and going through the full dataset one time just to count samples post-padding.

**Comment by mguillau** [comment]
Yes, indeed. The point is to simplify a call like:

`model.fit_generator( traingenerator, samples_per_epoch=traingenerator.samples_per_epoch, validation_data=valgenerator, nb_val_samples=valgenerator.samples_per_epoch, ... )`

with

`model.fit_generator( traingenerator, validation_data=valgenerator, ... )`

It's up to the user to provide this attribute with its generators or not (your code is one way to do it), but in any case `fit_generator` (and others) should be modified to look for it as well.

**Comment by carlthome** [comment]
@mguillau, what's the benefit of hiding `samples_per_epoch` in `traingenerator` though? There's a point to having a flat dependency injection graph for the API. `fit_generator` should clearly state upfront what it needs to function. Why should `traingenerator` have an attribute it doesn't actually need? (see [SOLID](https://en.wikipedia.org/wiki/SOLID_(object-oriented_design)))

**Comment by mguillau** [comment]
@carlthome, just to be clear, I am not suggesting to remove the `samples_per_epoch` argument, just to make it optional -- which is actually the title of this PR ;-).

Many generators already have this attribute in some form. `keras.preprocessing.ImageDataGenerator` is a good example: it needs it to sample a random permutation and to check the consistency of the provided input data and labels.

I agree that there's a risk for generators that, by chance, are using the `samples_per_epoch` attribute with completely different semantics. Just needs to be documented.

**Comment by carlthome** [comment]
You're not proposing to make it optional. You're proposing to hide it in an abstraction. There's no clear benefit.
